cf-custom-resources/lib/dns-cert-validator.js

```diff
@@ -161,19 +192,210 @@ const requestCertificate = async function (
+        promises.push(
+          validateDomain({
+            route53: appRoute53,
+            record: option.ResourceRecord,
+            action: action,
+            domainName: `${appName}.${domainName}`,
+          })
+        );
+        break;
+      case domainName:
+        promises.push(
+          validateDomain({
+            route53: appRoute53,
+            record: option.ResourceRecord,
+            action: action,
+            domainName: domainName,
+          })
+        );
+        break;
+    }
+  }
+  return Promise.all(promises);
+};
+
+const deleteHostedZoneRecords = async function (
+  options,
+  envName,
+  appName,
+  domainName,
+  envRoute53,
+  appRoute53,
+  acm,
+  envHostedZoneId
+) {
+  let isLegacyCert = false;
+  // Legacy cert only has two DomainValidationOptions:
+  // `${envName}.${appName}.${domainName}` and `*.${envName}.${appName}.${domainName}`
+  if (options.length <= 2) {
+    isLegacyCert = true;
+  }
+
+  let certWithDomainCount = await numOfGeneratedCertificates(
+    acm,
+    `${envName}.${appName}.${domainName}`
+  );
+  let isLastOne = false;
+  if (certWithDomainCount < 2) {
+    isLastOne = true;
+  }
+
+  const newOptions = [];
+  switch (`${isLegacyCert}|${isLastOne}`) {
+    case `true|true`:
+      // If it is a legacy cert and it is the last Copilot cert,
+      // we'll go ahead to remove the validation CNAME record in env hosted zone.
+      newOptions.push(...options);
+      break;
+    case `true|false`:
+      // If it is a legacy cert but it is not the last Copilot cert,
+      // we'll do nothing since the new Copilot cert needs validation CNAME record
+      // in the env hosted zone.
+      break;
+    case `false|true`:
+      // If it is not a legacy cert and it is the last Copilot cert,
+      // we'll remove all validation CNAME records in env/app/root hosted zone.
+      newOptions.push(...options);
+      break;
+    case `false|false`:
+      // If it is not a legacy cert and it is not the last Copilot cert,
+      // we'll remove validation CNAME records only for app and root hosted zone,
+      // since the legacy cert still needs the validation record in the env hosted zone.
+      for (const option of options) {
+        if (option.DomainName === `${envName}.${appName}.${domainName}`) {
+          continue;
+        }
+        newOptions.push(option);
+      }
+      break;
+  }
+  await updateHostedZoneRecords(
+    "DELETE",
+    newOptions,
+    envName,
+    appName,
+    domainName,
+    envRoute53,
+    appRoute53,
+    envHostedZoneId
+  );
+};
+
+// numOfGeneratedCertificates returns the number of Copilot generated certificates. Specifically it returns
+// the number of certificate with domain name as `${envName}.${appName}.${domainName}`, since if any of
+// the other certificate with this domain name still exists, we cannot remove the validation record in our
+// env hosted zone.
+const numOfGeneratedCertificates = async function (acm, defaultEnvDomain) {
```

Comment:
Let's remove the explanation of the env domain from this function to `deleteHostedZones` as that's the more appropriate place 🙏  and rename the parameter

```suggestion
// numOfGeneratedCertificates returns the number of Copilot generated certificates for a given domain name. 
const numOfGeneratedCertificates = async function (acm, domainName) {
```

For example (please feel free to edit for it to make more sense)
```javascript
// Deletes the validation records associated with the certificate.
// We don't want to delete a validation record if it's used by another certificate because
// the validation records are used to renew the certificate. 
// The legacy certificate (only `${envName}.${appName}.${domainName}`) and the new 
// certificate generated with the "alias" field share the same validation record. 
// Therefore, we check if there is more than one certificate using the record and only delete
// if there is no other certificate using the record.
const deleteHostedZoneRecords = ...
```